java-checks/src/main/java/org/sonar/java/checks/unused/UnusedPrivateMethodCheck.java

```diff
@@ -165,7 +168,14 @@ private static boolean isUnusedPrivate(Symbol symbol) {
   }
 
   private static boolean hasNoAnnotation(MethodTree methodTree) {
-    return methodTree.modifiers().annotations().isEmpty();
+    return methodTree.modifiers().annotations().isEmpty() && methodTree.parameters().stream().allMatch(UnusedPrivateMethodCheck::hasNoAnnotation);
```

Comment:
Okay I think this is not enough here if we want to avoid FPs in case of misconfigured projects. I would prefer that we try to avoid having a new rule generating potential FP, so we would privilege FN. It means that we need to be able to react correctly in absence of semantic... and so we have to rely on the tokens only.

Here is the fix I foresee instead:

```java
  private static boolean hasNoAnnotation(MethodTree methodTree) {
    return methodTree.modifiers().annotations().isEmpty()
      && methodTree.parameters().stream().noneMatch(UnusedPrivateMethodCheck::hasNoAllowedAnnotation);
  }

  private static boolean hasNoAllowedAnnotation(VariableTree variableTree) {
    List<AnnotationTree> annotations = variableTree.modifiers().annotations();
    return !annotations.isEmpty()
      && annotations.stream().anyMatch(UnusedPrivateMethodCheck::isAllowedAnnnotation);
  }

  private static boolean isAllowedAnnnotation(AnnotationTree annotation) {
    Type annotationSymbolType = annotation.symbolType();
    if (annotationSymbolType.is(PARAM_ANNOTATION_EXCEPTION)) {
      // semantic-based
      return true;
    }
    if (annotationSymbolType.isUnknown()) {
      // absence of semantic - rely on tokens to build the name
      TypeTree annotationType = annotation.annotationType();
      if (annotationType.is(Tree.Kind.IDENTIFIER)) {
        // might lead to FN in case of wrongly configured project - acceptable
        return "Observes".equals(((IdentifierTree) annotationType).name());
      }
      if (annotationType.is(Tree.Kind.MEMBER_SELECT)) {
        return PARAM_ANNOTATION_EXCEPTION.equals(ExpressionsHelper.concatenate((MemberSelectExpressionTree) annotationType));
      }
    }
    return false;
  }
```